File: interview/prob207.go
func canFinish(numCourses int, prerequisites [][]int) bool {
    
    m := len(prerequisites)
    if m == 0 {
        return true
    }

    // count prerequsites of each course
    inDegree := make([]int, numCourses)

    hmap := map[int][]int{}
    for _, pair := range prerequisites {
        course, pre := pair[0], pair[1]
        inDegree[course] += 1
        list, ok := hmap[pre]
        if ! ok {
            hmap[pre] = []int{course}
        }
        hmap[pre] = append(list, course)
    }

    queue := []int{}

    for i := 0; i < numCourses; i++ {
        if inDegree[i] == 0 {
            queue = append(queue, i) 
        }
    }

    remain := numCourses
    var course int
    for len(queue) > 0 {
        queue, course = queue[1:len(queue)], queue[0] 
        remain -= 1
        list, ok := hmap[course]
        if ok {
            for _, tmp := range list{
                inDegree[tmp] -= 1
                if inDegree[tmp] == 0 {
                    queue = append(queue, tmp)
                }
            }
        }

    }

    return remain == 0

}

File: interview/prob207.py
class Solution:
    def canFinish(self, numCourses: int, prerequisites: List[List[int]]) -> bool:
        
        from collections import defaultdict
        
        dep = defaultdict(set)
        
        for a, b in prerequisites:
            dep[a].add(b)  # a dependes on b
            
        courses = set(i for i in range(numCourses))
        
        while True:
            dependent = dep.keys()
            if not dependent:
                return True
            independent = courses - set(dependent)  # course you can take
            if not independent:
                return False
            
            courses.difference_update(independent)
            
            for key in independent:
                for key2 in dep:
                    if key in dep[key2]:
                        dep[key2].remove(key)
                
            toremove = set()
            for key in dep:
                if not dep[key]:
                    toremove.add(key)
                    
            for key in toremove:
                del dep[key]
